src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -1227,8 +1227,12 @@ bool StorageReplicatedMergeTree::checkTableStructure(const String & zookeeper_pr
     Coordination::Stat columns_stat;
     auto columns_from_zk = ColumnsDescription::parse(zookeeper->get(fs::path(zookeeper_prefix) / "columns", &columns_stat));
 
-    const ColumnsDescription & old_columns = metadata_snapshot->getColumns();
-    if (columns_from_zk == old_columns)
+    const auto & old_columns = metadata_snapshot->getColumns();
+
+    /// Replicated tables on different replicas must have exactly same column definitions
+    /// We cannot just compare column descriptions here because data types like SimpleAggregateFunction
+    /// may have different aggregate function in 1st argument but still compatible if 2nd argument is same.
```

Comment:
> data types like SimpleAggregateFunction may have different aggregate function in 1st argument but still compatible if 2nd argument is same

Why? This looks like the root cause, and probably we should fix SimpleAggregateFunction types comparison instead

Reply:
I don't think we should do that.  Refer to Alexey comment above when I changed comparison logic for `ColumnDescription`

> The code is wrong because it applies not only to Replicated tables,
> but we allow aggregate functions with compatible states for Distributed, Merge, and Buffer tables.

So basically as far as I understand, except for replicated table, we want `SimpleAggregateFunction` to be compatible (as long as 2nd arguments are same) in almost all other cases.

Reply:
> The code is wrong because it applies not only to Replicated tables

Well, I agree that the code was wrong but definitely not because of this :) 
The right way was to fix the `equals` method for `SimpleAggregateFunction`, not to compare type names as strings

> but we allow aggregate functions with compatible states for Distributed, Merge, and Buffer tables.

Why would we allow that? Or why would anyone use that? Looks like a way to shoot your leg.

> So basically as far as I understand, except for replicated table, we want SimpleAggregateFunction to be compatible (when 2nd arguments are same) in almost all other cases.

I don't know what Alexey wants :) 

> And this is a disaster if we don't find the issue fast enough.

What is the disaster exactly?

Reply:
> The right way was to fix the equals method for SimpleAggregateFunction, not to compare type names as strings

From implementation POV, `SimpleAggegateFunction` is not really a `IDataType` but rather an ordinary `IDataType` with some custom name and serialisation:

```cpp
  /// This is mutable to allow setting custom name and serialization on `const IDataType` post construction.
  mutable DataTypeCustomNamePtr custom_name;
  mutable SerializationPtr custom_serialization;
```

So I guess you want to compare `custom_name` if they're non-null? The scope will be broader than just to fix the issue in this PR. Waiting for final decision whether we should do this or not.

> Why would we allow that? Or why would anyone use that? Looks like a way to shoot your leg.

May be @alexey-milovidov can answer this question :)

> What is the disaster exactly?

Data inconsistency - some aggregated with `max`, and some aggregated with `anyLast`, for example.
Replicated merges have logic to check checksum and download to make part to be byte-identical, but we can't control which replica will download from which replica. In reality, we observe that both replicas download from the other to fix checksum mismatch. So in the end all replicas have mixed data.